File: server/tools/read_web_page.js
// Skill: fetches a web page and returns its text content, so Jarvis can
// summarize it or answer questions about it in conversation. (Unlike
// web_search, which just opens a results page for the user to read
// themselves.)
//
// The fetching and HTML-stripping now live in server/media.js, shared with
// Content Analysis and the research engine — there were two copies of the
// same stripHtml drifting apart, and media.js's version also drops nav,
// header and footer blocks, so what comes back is the article rather than
// the site's menu.
//
// The character cap stays low here on purpose. This result goes into a
// spoken conversation turn, where a huge wall of text costs tokens and slows
// the reply; Content Analysis asks for far more because it's building a
// digest to keep.
//
// A plain fetch gets nothing useful from a JS-rendered page (a real
// single-page app with no server-rendered HTML) — when that happens, this
// falls back to connectors/browser.js's renderPageHeadless(), a completely
// invisible (no window ever appears) real Chrome render, so "most lookups
// shouldn't pop a browser window" stays true even for those pages. Safe to
// import directly (browser.js is a leaf module — see its own header
// comment; no circular-import risk per root CLAUDE.md's invariant).

import { fetchArticle } from '../media.js';
import { renderPageHeadless } from '../connectors/browser.js';

const MAX_CHARS = 6000;
// Below this, a plain fetch's result is treated as "didn't really get the
// page" (an SPA's near-empty server-rendered shell, a JS-gated paywall
// stub, ...) rather than "the page is just short" — worth a real headless
// render instead of returning a near-nothing answer.
const THIN_RESULT_CHARS = 200;

export default {
  name: 'read_web_page',
  description:
    'Fetch a web page and read its text content, so you can summarize it or answer questions about it. ' +
    'Use this when the user gives you a URL, or after web_search if they want to know what a result actually says.',
  parameters: {
    type: 'object',
    properties: {
      url: { type: 'string', description: 'The full URL to read, e.g. "https://example.com/article".' },
    },
    required: ['url'],
  },
  async run({ url } = {}) {
    const result = await fetchArticle(url, { maxChars: MAX_CHARS });
    if (result.ok && result.text && result.text.trim().length >= THIN_RESULT_CHARS) {
      return { ok: true, url: result.url, title: result.title, text: result.text };
    }
    // Plain fetch either failed outright or came back too thin to be real
    // content — try a real, but still fully invisible, render before giving
    // up. Never pops a window either way.
    try {
      const rendered = await renderPageHeadless(url, { maxChars: MAX_CHARS });
      if (rendered.text && rendered.text.trim().length > 0) {
        return { ok: true, url: rendered.url, title: rendered.title, text: rendered.text };
      }
    } catch {
      // Fall through to whatever the plain fetch already told us — a
      // headless-render failure isn't a better error than the original one.
    }
    if (result.ok) return { ok: true, url: result.url, title: result.title, text: result.text };
    return { ok: false, error: result.error };
  },
};

File: backend/jarvis/tools/read_web_page.py
"""Read a web page the user already has a URL for, as plain text.

Deliberately does NOT open a visible browser window: a plain read is invisible
work. Three levels, and it only goes down one when the level above genuinely
came back with nothing: a plain fetch, then a headless render for a page that
builds itself in the browser, and — only as something to SAY, never something
this tool does on its own — the visible browser, which is for pages that have to
be clicked or typed into.

That last step is a sentence rather than a call on purpose. Opening a window on
someone's screen is a thing they should have asked for; a lookup that quietly
pops one up is the behaviour this ordering exists to prevent.
"""

from __future__ import annotations

import re

from ..capabilities import CapabilitySpec, Risk
from ..webtext import title_of, to_text
from ._http import get_text

MAX_CHARS = 20000
#: Below this, what came back is almost certainly a page shell rather than an
#: article — worth rendering properly before reporting there is nothing there.
THIN_CHARS = 600
def _run(url: str = "") -> dict:
    target = str(url or "").strip()
    if not re.match(r"^https?://", target, re.I):
        return {"ok": False, "error": "That doesn't look like a web address."}
    try:
        markup = get_text(target)
    except Exception as err:  # noqa: BLE001
        return {"ok": False, "error": f"Couldn't read that page ({err.__class__.__name__})."}
    text = to_text(markup)
    rendered = False
    if len(text) < THIN_CHARS:
        # Almost nothing came back. That is usually a page that builds itself in
        # the browser rather than a page with nothing on it, so it is worth
        # actually running — headlessly, invisibly. A plain fetch that came back
        # thin and an empty page look identical from here, and answering from
        # the shell of a page is worse than taking the extra second.
        from ..webrender import render

        attempt = render(target)
        if attempt.ok:
            fuller = to_text(attempt.html)
            if len(fuller) > len(text):
                markup, text, rendered = attempt.html, fuller, True

    if not text:
        if rendered:
            # It was really run, and still had nothing readable. What is left is
            # a page that needs interacting with — which is the browser
            # connector's job, and the user's call to make.
            return {"ok": False,
                    "error": "That page had no readable text even after running it. If it "
                             "needs signing in or clicking through, say so and I can open "
                             "it in a browser window you can watch."}
        return {"ok": False,
                "error": "That page had no readable text — it may need a real browser."}
    return {"ok": True, "url": target, "title": title_of(markup), "rendered": rendered,
            "truncated": len(text) > MAX_CHARS, "text": text[:MAX_CHARS]}


SPEC = CapabilitySpec(
    id="builtin.read_web_page",
    name="read_web_page",
    description=("Read the text of a web page you already have the address for. Use this to "
                 "look something up quietly rather than opening a browser window."),
    input_schema={"type": "object", "properties": {
        "url": {"type": "string", "description": "The full web address, starting with http."}},
        "required": ["url"]},
    risk=Risk.LOW,
    handler=_run,
    timeout_s=25.0,
)
